[CORE-12697] Implement support for license format v1

IoannisRP opened the pull request:
Implements: [CORE-12697](https://redpandadata.atlassian.net/browse/CORE-12697)

Implement license v1 format. This includes:
- Change type of `type` to a string
- Allow optional list of products
- Allow additional parameters

If `products` is included in the license, it will also be returned by the `get_license` admin endpoint.

Note on backport: This is a feature aimed at 25.2.x. However it is merged after the 25.2.x branch has been cut.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

* none


[CORE-12697]: https://redpandadata.atlassian.net/browse/CORE-12697?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

IoannisRP committed:
security: Add utility function in license test

The new utility function will be used in the following commits.

Replaced boost check with runtime error to make it clearer
that if this is encountered it is an issue of the test.

IoannisRP commented:
/dt

IoannisRP commented:
/dt

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/a3e8a779da5557f7a02f4204873fbb5830f80b8b..af3e2e04b024e4bb955bed0e862ee41a20bdb812):
- Updated serialization of license type to be backwards compatible.
- Removed test for additional parameters.
- Replaced placeholder values with actual values in v1 tests.

IoannisRP committed:
rp/admin: Update licensing description

With licensing format v1, any value can be passed into the type
variable. Remove explicit values from description.

michael-redpanda commented on src/v/security/tests/license_test.cc:
```diff
@@ -18,6 +18,24 @@
 
 using namespace std::chrono_literals;
 
+namespace {
```
pipedream: convert this test into google test (don't do it in this PR, but would be nice....)

IoannisRP replied:
nice follow up when this PR is done and merged :+1:

michael-redpanda commented on src/v/security/license.h:
```diff
@@ -72,7 +72,7 @@ struct license
     /// Expected encoded contents
     uint8_t format_version;
-    license_type type;
+    ss::sstring type;
```
yeah changing this is going to be bad and need updates to the serde::version and serde::compat_version fields.

Also may need an entry in the feature table to ensure the cluster has been updated before this change is handled.

@BenPope @pgellert  - do ya'll know of any examples in the not-too-distant past of changing types and how that impacts the serde version fields?

michael-redpanda replied:
I see you changed the compat verisons in a later commit - i think by convention you should change the verisons in the commit where you're modifying the envelope

michael-redpanda commented on src/v/security/license.h:
```diff
@@ -75,10 +75,12 @@ struct license
     auto serde_fields() {
-        return std::tie(format_version, type, organization, expiry, checksum);
+        return std::tie(
+          format_version, type, organization, expiry, products, checksum);
```
adding fields requires updates to the serde::version number

BenPope commented on src/v/security/license.cc:
```diff
@@ -215,13 +231,17 @@ fmt::formatter<security::license, char, void>::format<
-      r.organization,
-      r.type,
-      r.expiry.count());
+    std::vector<std::string> fields{
+      fmt::format("Version: {}", r.format_version),
+      fmt::format("Organization: {}", r.organization),
+      fmt::format("Type: {}", r.type),
+      fmt::format("Expiry(epoch): {}", r.expiry.count()),
+    };
+    if (!r.products.empty()) {
+        fields.push_back(
+          fmt::format("Product(s): {}", fmt::join(r.products, ",")));
+    }
+    return fmt::format_to(ctx.out(), "[{}]", fmt::join(fields, ", "));
```
nitpick: I don't think we need all of the temporaries:
```suggestion
    return fmt::format_to(
      ctx.out(),
      "[Version: {0}, Organization: {1}, Type: {2}, Expiry(epoch): {3}{4}]",
      r.format_version,
      r.organization,
      r.type,
      r.expiry.count(),
      r.products.empty()
        ? ""
        : fmt::format(", Product(s): {}", fmt::join(r.products, ",")));
);

BenPope commented on src/v/security/license.h:
```diff
@@ -75,10 +75,12 @@ struct license
 
     auto serde_fields() {
-        return std::tie(format_version, type, organization, expiry, checksum);
+        return std::tie(
+          format_version, type, organization, expiry, products, checksum);
```
Doesn't changing the order of these break compatibility?
```suggestion
        return std::tie(
          format_version, type, organization, expiry, checksum, products);

IoannisRP replied:
i think the compatibility got broken when the type of `type` got changed from `enum` -> `string`. No?

I'll address this as part of the rest of the response.

michael-redpanda replied:
Yeah I think we should go simple here - add a new field like "type_string" and depending on the version of the license either reference or don't reference it.

IoannisRP replied:
yeah, this is the approach i am going with. :+1:

BenPope commented on src/v/security/license.h:
```diff
@@ -67,7 +67,7 @@ inline std::ostream& operator<<(std::ostream& os, license_type lt) {
 
 struct license
-  : serde::envelope<license, serde::version<1>, serde::compat_version<0>> {
+  : serde::envelope<license, serde::version<2>, serde::compat_version<2>> {
```
Can a v0 or v1 license be read from the controller log now?

BenPope replied:
> Can a v0 or v1 license be read from the controller log now?

```
ERROR 2025-07-10 14:24:49,750 [shard 0:main] main - application.cc:3405 - Exception decoding feature table snapshot: serde::serde_exception (read_header.h:82 - read security::license: version=1 < security::license::compat_version=2)
ERROR 2025-07-10 14:24:49,750 [shard 0:main] assert - Assert failure: (src/v/redpanda/application.cc:3407) 'false' Snapshot decode failed
```

IoannisRP replied:
fixed

BenPope reviewed (dismissed):
It's not obvious to me that reading an old licence from the controller log can work due to the bump of compat version to 2.

I'd prefer adding fields to the end of `serde_fields` rather than changing the types and orders.

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/a0417119ac3c7771a3991dac529641a1c77547af..75d083f1b2d6234e236f60503662d3ebc40a7d02):
- Reworked changes to be serde-backwards-compatible
- Added stronger railguards around license versioning

IoannisRP committed:
rp/admin: Update get license response

Add products to get_license response, if they are not empty

BenPope commented on src/v/security/license.h:
```diff
@@ -67,18 +67,24 @@ inline std::ostream& operator<<(std::ostream& os, license_type lt) {
     /// Expected encoded contents
     uint8_t format_version;
-    license_type type;
+    // This variable should not be used directly. Use get_type accessor, instead
+    license_type _type;
```
These should be moved to the implementation file:
```c++
ss::sstring license_type_to_string(license_type type);

inline std::ostream& operator<<(std::ostream& os, license_type lt) {
    os << license_type_to_string(lt);
    return os;
}
```

BenPope commented on src/v/security/license.cc:
```diff
@@ -179,6 +190,14 @@ static void parse_data_section(license& lc, const json::Document& doc) {
+            std::ranges::transform(
+              products, std::back_inserter(lc.products), [](const auto& e) {
+                  return e.GetString();
+              });
```
nitpick:
```suggestion
            lc.products = it_products->value.GetArray()
                          | std::views::transform(&json::Value::GetString)
                          | std::ranges::to<decltype(lc.products)>();
```
There's a helper floating around:
```c++
inline std::string_view as_string_view(const json::Value& v) {
    vassert(v.IsString(), "You must check that v is a string;");
    return {v.GetString(), v.GetStringLength()};
}
```
Which avoids the `strlen`, but I'm not sure it's worth it here.

IoannisRP replied:
I went with your suggestion, without the helper.

IoannisRP committed:
security: Move static functions/variables to unamed namespaces

The is a pure refactor commit. No changes in behavor.

pgellert commented on src/v/features/feature_table.cc:
```diff
@@ -306,7 +306,7 @@ make_builtin_trial_license(security::license::clock::time_point start_time) {
 
     return security::license{
       .format_version = 0,
-      .type = security::license_type::free_trial,
```
Should this now be a `format_version=1` license?

IoannisRP replied:
my understanding is that we want to delay the rollout of updated versions for a while, but we want to ability to generate them to be in place.

IoannisRP replied:
As long as version `0` is supported, i think it's fine. These are cleaner, as well.

pgellert commented on src/v/security/license.h:
```diff
@@ -67,18 +67,24 @@ inline std::ostream& operator<<(std::ostream& os, license_type lt) {
     /// Expected encoded contents
     uint8_t format_version;
-    license_type type;
+    // This variable should not be used directly. Use get_type accessor, instead
+    license_type _type;
```
`_type` and `_type_str` are currently "suggested private". We should make them truly private to ensure that it can't be misused. We should also guarantee that only valid licenses can be constructed (eg. add 2 constructors/factory methods for the two format versions)

IoannisRP replied:
It would take a full `pimple` to properly hide them and have clean accessors, because of how the parsing is structured (the parsing also needs to know about json, so it has to live entirely in the `.cc`). I don't think it is worth the effort.

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/37bf0fb295420702bb870763346a23d3698fae0c..edf82bd9d668f1aaf20d277d5310e46b8c122cd6): 
- addressed feedback
- refactored all static functions/vars into unamed namespaces
- updated tests to their respective licenses

michael-redpanda commented on src/v/security/license.cc:
```diff
@@ -158,12 +164,17 @@ void parse_data_section(license& lc, const json::Document& doc) {
             throw license_invalid_exception(
               "Unsupported entry for type. Expected int value");
         }
+        if (has_products_field) {
+            throw license_invalid_exception("Unsupported products field");
+        }
```
Nothing wrong here, but do we want to define multiple schemas above, one for v0 and one for v1 so we don't have to do this check?  cc @BenPope

BenPope replied:
> Nothing wrong here, but do we want to define multiple schemas above, one for v0 and one for v1 so we don't have to do this check? cc @BenPope

I don't have strong feelings here.

michael-redpanda commented on src/v/security/license.h:
```diff
@@ -72,12 +72,19 @@ struct license
     ss::sstring checksum;
     // This variable should not be used directly. Use get_type accessor, instead
     ss::sstring _type_str;
+    std::vector<ss::sstring> products;
```
suggestion: chunked_vector?

BenPope replied:
> suggestion: chunked_vector?

I think 8193 products is probably overkill.

michael-redpanda commented on src/v/security/license.cc:
```diff
@@ -135,7 +135,7 @@ const std::string license_data_validator_schema = R"(
         "expiry"
     ],
-    "additionalProperties": false
+    "additionalProperties": true
```
I think this change warrants defining a separate schema for v1

michael-redpanda reviewed (commented):
nice - outstanding quesiton is whether or not to define a v1 schema (like there is a v0 schema)

pgellert commented on src/v/security/tests/license_test.cc:
```diff
@@ -132,14 +132,14 @@ BOOST_AUTO_TEST_CASE(test_license_valid_content_v1_products) {
     BOOST_CHECK_EQUAL(
       license.checksum,
-      "65d9629c13bb13f12f4a2a973677b16700dd0a276c0759d29e78a73689fa91f2");
+      "0937a2d8e4437a63373c1c1cb0f5f62c5cae9366fea1b00467b4c4eaab8ca4cf");
```
Should this be in a different commit?

IoannisRP replied:
correct. Was squashed to the wrong commit

pgellert commented on src/v/security/license.cc:
```diff
@@ -141,11 +148,37 @@ void parse_data_section(license& lc, const json::Document& doc) {
     if (lc.format_version < 0) {
         throw license_invalid_exception("Invalid format_version, is < 0");
     }
+    if (lc.format_version > 1) {
+        throw license_invalid_exception("Unsupported format_version");
+    }
```
nit: should this be part of the commit 75ad9233ca6b86fa05f2ca5d6c05e3bb6f4bfb6a? Or that commit merged into this one?

IoannisRP replied:
the whole thing will be reworked.

pgellert commented on tests/rptest/tests/cluster_features_test.py:
```diff
@@ -239,6 +239,35 @@ def obtain_configured_license():
+            'type': 'testing_license',
+            'products': ['some_prod', 'some_other_prod'],
+            'expires': 0,
+            'sha256': 'PLACEHOLDER'
```
Will this need updating? Or should we just ignore the `sha` in the equality check on L269?

IoannisRP replied:
it needs updating. I am just waiting for the test to fail to verify that everything else is correct about this test.

IoannisRP committed:
security: Refactor code for parsing license json

The code responsible for parsing the raw license into a license
object has been refactored in a way that is scalable to directly
support more format versions.

Other changes:
- Remove redundant check for negative format version.
By the time this check is performed, version has already been
cast into a uint8_t. On top of it, all non-supported versions
are now covered by the version switch statement.
- Fixed and extended error messages in parsing.
- Zero-initialize license to ensure no field is left uninitialized.

IoannisRP committed:
security: Changed type of license entry

Changed the type of 'type' from int to string.

This is a part of multiple changes to move from license format V0 -> V1:
- Change type of 'type' to a string
- Allow optional list of products
- Allow additional parameters

Change 2/4

IoannisRP committed:
security: Add support for licensing format version 1

This commit only introduces the parsing functions needed for
support of v1. It is currently identical to v0. Functional
changes are introduced in the following commits.

This is a part of multiple changes to move from license format V0 -> V1:
- Change type of 'type' to a string
- Allow optional list of products
- Allow additional parameters

Change 1/4

IoannisRP committed:
security: Add products entry in license

Add optional list of products in license json

This is a part of multiple changes to move from license format V0 -> V1:
- Change type of 'type' to a string
- Allow optional list of products
- Allow additional parameters

Change 3/4

IoannisRP committed:
security: Enable support for arbitrary parameters in license

There is no unit testing of this functionality, as the license generator
doesn't support generation of arbitrary parameters. Tested locally
with a hand-generated license.

This is a part of multiple changes to move from license format V0 -> V1:
- Change type of 'type' to a string
- Allow optional list of products
- Allow additional parameters

Change 4/4

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/edf82bd9d668f1aaf20d277d5310e46b8c122cd6..7ab35f6eb5dbd168073c8fcd76fe7bac5b0c795e):
- reworked license versioning. Each version has a separate schema/parsing path.
- added correct values to dt test.

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/7ab35f6eb5dbd168073c8fcd76fe7bac5b0c795e..cecad074a789bb975db8ac7cd80c39fd4a80fdc0):
- Removed redundant check. Explanation in the commit message.

michael-redpanda commented on src/v/security/license.cc:
```diff
@@ -126,28 +126,61 @@ const std::string license_data_validator_schema = R"(
-          "License data section failed to match schema");
-    }
+struct license_data_parser {
+    using data_parser = void (*)(license& lc, const json::Document& doc);
```
can this be a `std::function` rather than a C-style function pointer?

IoannisRP replied:
it could, but it feels like an overkill.
I don't see this needing any of the benefits of a `std::function` (i.e. functor objects), while it would suffer from its drawbacks (it's a heavy-weight object with an extra indirection due to type erasure).

If anything i was considering going templated to avoid the non-inlining that this approach has (i don't think it's worth the extra code complexity of that, though).

BenPope replied:
A function pointer is fine here.

michael-redpanda replied:
I mean you are both correct but I disagree :P 

all good though ;)

michael-redpanda commented on src/v/security/license.cc:
```diff
@@ -167,10 +167,6 @@ void parse_data_section(license& lc, const json::Document& doc) {
-    if (lc.format_version < 0) {
-        throw license_invalid_exception("Invalid format_version, is < 0");
-    }
-
```
chore - should this just be in the previous commit?

michael-redpanda commented on src/v/security/license.cc:
```diff
@@ -137,7 +137,7 @@ const char* const license_data_validator_schema_v1 = R"(
         },
         "type": {
-            "type": "number"
+            "type": "string"
```
I'd hold off with putting in the schema until after the changes in this commit are done.  Effectively change the order of the commits.  Establish how the license is stored on Redpanda and then write the parsing logic.

IoannisRP replied:
I am not sure i see the benefit. The way i see it, the` make_license` function is a named constructor. The parsing functions are free functions, but that is an implementation detail. Why split the data structure from how it is constructed?

michael-redpanda commented on src/v/security/license.cc:
```diff
@@ -155,7 +155,7 @@ const char* const license_data_validator_schema_v1 = R"(
         "expiry"
     ],
-    "additionalProperties": false
+    "additionalProperties": true
```
Should this be in the commit that creates the schema?

michael-redpanda reviewed (commented):
I think the commit ordering can be improved here.  I really like the license validator abstraction.

I think you want to:

1. Make the necessary changes for the in memory data structure
2. Make the changes that you made to make license validation scalable
3. Add in v1 in its entirety

Otherwise you're making changes in one commit and then changing them in another - it's a little confusing to review.

IoannisRP commented:
@michael-redpanda 

> Otherwise you're making changes in one commit and then changing them in another

While it is true, i think it clearly shows the 3 orthogonal updates from v0 -> v1.
Also it has the benefit of every change being followed by its respective test (where applicable).

Otherwise, we have the data structure changes without any tests and then the whole v1 parsing with "all" (i.e. 2) tests.

Given the scope of this changes (it's not much code in total), I could squash all the */4 commits, so all the data structure, the construction logic and the tests stay together, if you thing it would improve the history of it.

IoannisRP commented:
in [force-push](https://github.com/redpanda-data/redpanda/compare/cecad074a789bb975db8ac7cd80c39fd4a80fdc0..67bcd54f9cd655ef8aa1fbb71df4030a97752145): history modification. No code changes

BenPope commented on src/v/security/license.cc:
```diff
@@ -99,7 +99,7 @@ license_components parse_license(std::string_view license) {
 }
 
-const std::string license_data_validator_schema = R"(
+const char* const license_data_validator_schema_v0 = R"(
```
`json::validator` is constructed with a `const std::string&`. I don't know the type was changed here.

IoannisRP replied:
having a static std::string here results in a constant duplication of the string data into heap.
A `const char*` removes the duplication, at a cost of runtime allocation. Optimally, this should be a `string_view`, but with the current interface of `json::validator` it would need a conversion, regardless.

Happy to have a followup to change the `json::validator` to a `string_view` as it doesn't really need a `std::string`.

BenPope replied:
> having a static std::string here results in a constant duplication of the string data into heap. A `const char*` removes the duplication, at a cost of runtime allocation.

I don't understand this comment. There will be an allocation at static initialisation time, either way. With `const char *` there will be an additional allocation each time the schema is parsed.

IoannisRP replied:
With static `std::string`, you have an allocation at initialization which stays alive for the full duration of this application.
With static `const char*`, you have an allocation when the schema validation happens and is released afterwards.

Note that rapidjson is operating using its own string types. Passing a `std::string` just copies it as the first thing, using `std::string::c_str`. My main argument here is `json::validator` shouldn't be written in terms of `std::string`. Everything else falls out of that.

michael-redpanda replied:
any more follow up here?  @IoannisRP I think I catch what you mean... any other questions @BenPope ?

BenPope commented on src/v/security/license.cc:
```diff
@@ -126,28 +126,57 @@ const std::string license_data_validator_schema = R"(
+    const auto it_version = doc.FindMember("version");
+    if (it_version == doc.MemberEnd()) {
+        throw license_invalid_exception(
+          "Missing required paramter 'format_version'");
```
```suggestion
        throw license_invalid_exception("Missing required parameter 'version'");
```

BenPope commented on src/v/security/license.cc:
```diff
@@ -126,28 +126,57 @@ const std::string license_data_validator_schema = R"(
+          .parser = parse_data_section_v0};
+    default:
+        throw license_invalid_exception(
+          fmt::format("Unsupported format_version {}", version));
```
```suggestion
          fmt::format("Unsupported version {}", version));
```

IoannisRP replied:
Good call. I was mindlessly copying what was there before. Will replace all `format_version` -> `version` in the errors.

BenPope commented on src/v/security/license.cc:
```diff
@@ -126,28 +126,57 @@ const std::string license_data_validator_schema = R"(
+    json::validator license_data_validator(ldf.schema);
+    if (!doc.Accept(license_data_validator.schema_validator)) {
+        throw license_malformed_exception(
+          "License data section failed to match schema");
```
```suggestion
        throw license_malformed_exception(ssx::sformat(
          "License data section failed to match schema: {} at offset {}",
          rapidjson::GetParseError_En(doc.GetParseError()),
          doc.GetErrorOffset()));
```

IoannisRP replied:
The error doesn't come from the doc. It comes from the validator.
Updated the relevant error message to include the schema validation error msg.

BenPope commented on src/v/security/license.cc:
```diff
@@ -148,12 +175,29 @@ void parse_data_section_v0(license& lc, const json::Document& doc) {
+    lc.organization = doc.FindMember("org")->value.GetString();
+    if (lc.organization == "") {
+        throw license_invalid_exception("Cannot have empty string for org");
+    }
```
nitpick: This is common with v0.

IoannisRP replied:
true, but i thought it's better to keep the 2 versions completely separate instead of having a 'common' and a 'diverging' section in the parsers. And it didn't feel worth the extra code to add a function to parse these two variables. My suggestion is that when/if we have yet another license version, we can extra common logic into whatever emerging pattern there is.

BenPope commented on src/v/redpanda/admin/server.cc:
```diff
@@ -2463,6 +2463,9 @@ void admin_server::register_features_routes() {
               lc.type = license->get_type();
               lc.expires = license->expiry.count();
               lc.sha256 = license->checksum;
+              if (!license->products.empty()) {
+                  lc.products = license->products;
+              }
```
I don't think the empty check is useful.
```suggestion
              lc.products = license->products;
```

I don't know why only a subset of the license is populated if the request is authenticated, and whether `products` should be in that conditional.

IoannisRP replied:
Auth changes added here #25224.
Based on the limited information, i don't think we need to move products into the "redacted" section.

IoannisRP replied:
Removing the `empty` check. On closer inspection, it is redundant here.

IoannisRP commented:
/dt

BenPope commented on src/v/security/license.cc:
```diff
@@ -126,28 +126,60 @@ const std::string license_data_validator_schema = R"(
+    json::validator license_data_validator(ldf.schema);
+    if (!doc.Accept(license_data_validator.schema_validator)) {
+        rapidjson::StringBuffer sb;
+        rapidjson::Writer<rapidjson::StringBuffer> writer(sb);
```
These version return null on alloc failure:
```suggestion
        json::StringBuffer sb;
        json::Writer<json::StringBuffer> writer(sb);
```

BenPope commented on src/v/security/license.cc:
```diff
@@ -126,28 +126,60 @@ const std::string license_data_validator_schema = R"(
+          .parser = parse_data_section_v0};
+    default:
+        throw license_invalid_exception(
+          fmt::format("Unsupported version {}", version));
```
nitpick:
```suggestion
          ss::format("Unsupported version {}", version));

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/14a19e2169860bf82cab02afb2749a92c3244062..59231c9e8af83d693020e154741b6c68f07bc2c6):
- move from `rapidjson` -> `json` for stringbuffer
- changed `fmt::format` -> `ss::format`

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--9591c018-1683-4aee-8654-b54ddb6ac390-->
<!--TEST RESULTS-->
<!--de83f284-4910-4c3e-81a1-e558382d236e-->
<!--TEST RESULTS-->
<!--23d6f11d-30b3-4040-b8b8-143d285b00cd-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69008">build#69008</a></summary>


| test_class                               | test_method                              | test_arguments                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                     |
|:-----------------------------------------|:-----------------------------------------|:----------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodePostRestartProbeTest                 | post_restart_probe_test                  |                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69008#01980f01-e501-416c-938f-e6349aa642e5 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS  |
| PrefixTruncateRecoveryTest               | test_prefix_truncate_recovery            | {"acks": -1, "start_empty": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69008#01980ee4-8a35-4be4-9cf0-d4ed7b588d38 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS |
| src/v/raft/tests/basic_raft_fixture_test | src/v/raft/tests/basic_raft_fixture_test |                                   | unit        | https://buildkite.com/redpanda/redpanda/builds/69008#01980ec7-2e4a-48aa-8073-032aeeaca2d9 | FAIL          | 0/1      |                                                                                                                                                            |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69115">build#69115</a></summary>


| test_class                  | test_method                             | test_arguments                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason   |
|:----------------------------|:----------------------------------------|:---------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|
| ShadowIndexingWhileBusyTest | test_create_or_delete_topics_while_busy | {"cloud_storage_type": 2, "short_retention": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69115#01981491-a301-437a-9ddf-f002009b581e | FLAKY         | 14/21    |          |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69202">build#69202</a></summary>


| test_class                               | test_method                         | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------------------------------|:------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageChunkReadTest                | test_read_chunks                    |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69202#0198196d-20ac-4253-bffa-7fd07090e8e6 | FLAKY         | 20/21    | upstream reliability is '96.95550351288055'. current run reliability is '95.23809523809523'. drift is 1.71741 and the allowed drift is set to 50. The test should PASS |
| ControllerLogLimitPartitionBalancerTests | test_partition_balancer_with_limits |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69202#01981970-3288-44b5-950d-22f7c70e0c93 | FAIL          | 0/1      |                                                                                                                                                                        |
</details>
<!--/body-->

michael-redpanda reviewed (changes requested):
Looks good - but I do think we need an upgrade test (25.1 with license -> upgrade to 25.2)

BenPope commented on src/v/security/license.cc:
```diff
@@ -185,7 +185,7 @@ void parse_data_section_v1(license& lc, const json::Document& doc) {
         throw license_invalid_exception("Cannot have empty string for org");
     }
-    lc.type = integer_to_license_type(doc.FindMember("type")->value.GetInt());
+    lc._type_str = doc.FindMember("type")->value.GetString();
```
`lc._type` should be set where possible, for backwards compatibility (if it's re-serialized)

BenPope replied:
i.e. add a test to ensure that v0 enterprise -> v1 -> v0 is still enterprise.

IoannisRP replied:
while i agree that uninitialized variables are bad (and fixed), i am not sure what situation you are referring to.
When we set the license back to a V0 one, the type will be re-set while parsing the license.

I added a test that i think does what you ask for. Let me know if you were asking for something else.

BenPope replied:
Oh right, license version 1 is rejected by older brokers, so having compatibility of the serde format is not required.

IoannisRP committed:
tests/rptest: Add tests to validate licensing format across versions

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/59231c9e8af83d693020e154741b6c68f07bc2c6..b50ff75d5e1f2e1667c12fff6f59857dc24a3ac5):
- Zero initialize `license` object
- Add extra tests

Feediver1 reviewed (dismissed):
descriptions look good for docs

BenPope commented on tests/rptest/tests/license_upgrade_test.py:
```diff
@@ -68,3 +69,79 @@ def license_loaded_ok():
+        assert res.status_code == 200, \
+                f"Expected status_code {200} but got {res.status_code} instead.\nContent: {res.content}"
+
+        # Update all nodes to newest version
+        self.installer.install(self.redpanda.nodes, (25, 2))
+        self.redpanda.restart_nodes(self.redpanda.nodes)
+        _ = wait_for_num_versions(self.redpanda, 1)
```
Is it worth testing what happens when a single node is upgraded, becomes the controller, and receives a v1 license? Should it be accepted? What do we expect from the other nodes?

michael-redpanda replied:
I think that's a very good idea - effectively let's make sure things don't come crashing down

IoannisRP replied:
added a new test with a partial upgraded cluster, with the controller being the newer node.

Also tested locally with the controller being on the older version. When sending a v1-formatted license on the newer version node, it forwards it to the controller and it fails there, as expected. But I don't think its worth the space for that to be part of the test.

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/b50ff75d5e1f2e1667c12fff6f59857dc24a3ac5..b70d676fd12b9f7fa7a82acad4adca50fc5cbb56):
- added extra versioning test

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/69202-->
<!--3fba34be-b875-490b-aa03-89c2a4f93ace-->
<!--title-->
## Retry command for [Build#69202](https://buildkite.com/redpanda/redpanda/builds/69202)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/controller_log_limiting_test.py::ControllerLogLimitPartitionBalancerTests.test_partition_balancer_with_limits
```
<!--/body-->

vbotbuildovich commented:
/backport v25.2.x
